pkg/controller/gitopscluster/gitopscluster_controller.go

```diff
@@ -767,3 +769,43 @@ func (r *ReconcileGitOpsCluster) CreateManagedClusterSecretInArgo(argoNamespace
 
 	return nil
 }
+
+func unionSecretData(newSecret, existingSecret *v1.Secret) *v1.Secret {
+	newLabels := newSecret.GetLabels()
+	existingLabels := existingSecret.GetLabels()
+
+	// union of labels
+	for key, val := range existingLabels {
+		newLabels[key] = val
+	}
+
+	newSecret.SetLabels(newLabels)
+
+	// union of annotations (except for kubectl.kubernetes.io/last-applied-configuration)
+	newAnnotations := newSecret.GetAnnotations()
+
+	existingAnnotations := existingSecret.GetAnnotations()
+	if existingAnnotations != nil {
+		newAnnotations = make(map[string]string)
+	}
+
+	for key, val := range existingAnnotations {
+		if key != "kubectl.kubernetes.io/last-applied-configuration" {
+			newAnnotations[key] = val
+		}
+	}
+
+	newSecret.SetAnnotations(newAnnotations)
+
+	// union of data
+	newData := newSecret.StringData
+	existingData := existingSecret.Data // api never returns stringData as the field is write-only
+
+	for key, val := range existingData {
+		newData[key] = string(val[:])
+	}
+
```

Comment:
To merge data from existing secret to the new secret,

1. always use the data.config, data.name, data.server in the new secret.
2. merge other key/values from existing secret to the new one

Reply:
Resolved by only adding key,values that do not already appear in the newSecret data.